---
title: "Genomic annotation by chromosomal location"
output:
  html_document:
    fig_caption: yes
    toc: yes
vignette: >
  %\VignetteEngine{knitr::rmarkdown} 
  %\VignetteIndexEntry{Genomic annotation by chromosomal location} 
  \usepackage[utf8]{inputenc}
---


<style type="text/css">
body {
    max: 800px;
    margin: 1em auto;
    max-width: 1200px;
}
hr {
  background-color: black;
  height:10px;
}
code < 
</style>




## How to use


```{r load-package}
suppressPackageStartupMessages(require(ProbeAnnotator))
require(org.Hs.eg.db)
require(TxDb.Hsapiens.UCSC.hg19.knownGene)
```


### Parameters

* `sep_intra` A character vector, by default `";"`, giving the seperator when concatenating multiple record from a gene. Take the _alias_ attribute of a gene. For instance, the gene HOXA1 (EntrezID:3198) has three other aliases BSAS, HOX1 and HOX1F. With default value, the genes alias will be `"BSAS;HOX1;HOX1F;HOXA1"`. Setting `sep_intra="|"` will give `"BSAS;HOX1;HOX1F;HOXA1"`. For example: 
```{r, eval=FALSE}
    probe_name  entrez_id   loctype start  end GENENAME SYMBOL  ALIAS
1   "X" 3198    gene 27132614 27135625 homeobox A1  HOXA1 BSAS;HOX1;HOX1F;HOXA1
```

* `sep_inter` A character vector, by default `"\\"`, giving the seperator when concatenating multiple genes in one probe (or row). Take for example a probe X mapping to two genes HOXA1 and HOXA2, the full record for this gene will be:
```{r, eval=FALSE}
    probe_name  entrez_id   loctype start  end GENENAME SYMBOL  ALIAS
1   "X" 3198\\3199    gene\\extended 27132614\\27139973 27142394\\27142394 homeobox A1\\homeobox A2  HOXA1\\HOXA2 BSAS;HOX1;HOX1F;HOXA1\\HOX1K;MCOHI
```


## Examples

-------

### Short Example

example of the MGMT gene on chromosome 10. The MGMT coordinates for the hg19 location is 27132612-2713562 on the sense strand.

We create three fake probe location, for three different MGMT `loctype`. Start site for these fake probes are:

* `"gene"` at 131263500 ($\text{probe} \in (start_{MGMT},end_{MGMT})$)
* `"promotor"` at 131264960 ($\text{probe} \in (start_{MGMT}\pm 15000)$))
* `"extended"` at 131265460 ($\text{probe} \in (start_{MGMT}-3000,start_{MGMT})$)

#### Example with GRanges object

```{r}
start = c(131263500, 131264960, 131265460)
probeID = sprintf("probe_%d",1:3)

## Using x in GRanges format
gr = GRanges(seqnames = "chr10",  
             strand = "+",
             ranges = IRanges(start = c(131263500, 131264960, 131265460),
                              width = 20),
             ID = sprintf("probe_%d",1:3))
annot_gr = annotateByLocation(x = gr, mapType = "EXON", verbose = FALSE)
print(gr)
print(annot_gr)
```


#### Example with GRanges object

```{r}
df = data.frame(chr = "chr10", 
                strand = "+", 
                start =  c(131263500, 131264960, 131265460), 
                end =  c(131263500, 131264960, 131265460)+19, 
                ID = sprintf("probe_%d",1:3))
annot_df = annotateByLocation(x = df, mapType = "EXON", verbose = FALSE)
print(df)
print(annot_df)
```

